cms/djangoapps/contentstore/outlines_backfill.py

```diff
@@ -0,0 +1,138 @@
+        return str(self.id)
+
+    @classmethod
+    def get_course_outline_ids(cls):
+        """
+        Returns all the CourseOverview object ids.
+        """
+        return cls.objects.values_list('id', flat=True)
+
+
+def backfill_course_outlines_subset(modeladmin, request, queryset):
+    """
+    Create a celery task to backfill a single course outline for each passed-in course key.
+
+    If the number of passed-in course keys is above a threshold, then instead create a celery task which
+    will then create a celery task to backfill a single course outline for each passed-in course key.
+    """
+    all_course_keys_qs = queryset.values_list('id', flat=True)
+    if len(all_course_keys_qs) >= TASK_TO_LAUNCH_TASKS_THRESHOLD:
+        # Create the celery task-creating task.
+        backfills = len(all_course_keys_qs)
+        update_multiple_outlines_from_modulestore_task.delay([str(course_key) for course_key in all_course_keys_qs])
+        log.info("Queuing task to create outline creation tasks for all (%s) courses.", backfills)
+    else:
+        # Create a separate celery task for each course outline requested.
+        backfills = 0
+        for course_key in all_course_keys_qs:
+            if key_supports_outlines(course_key):
+                log.info("Queuing outline creation for %s", course_key)
+                update_outline_from_modulestore_task.delay(str(course_key))
+                backfills += 1
+            else:
+                log.info("Outlines not supported for %s - skipping", course_key)
+    plural_suffix = 's' if backfills > 1 else ''
+    modeladmin.message_user(
+        request,
+        f"Successfully requested {backfills} course outline{plural_suffix} to be backfilled."
+    )
+backfill_course_outlines_subset.short_description = "Backfill selected course outlines"
+
+
+def backfill_course_outlines_all(modeladmin, request, queryset):
+    """
+    Custom admin action which backfills *all* the course outlines - no matter which CourseOverviews are selected.
+    This selection of all CourseOvervies occurs in the overridden changelist_view method below.
+    """
+    backfill_course_outlines_subset(modeladmin, request, queryset)
+backfill_course_outlines_all.short_description = "Backfill *all* course outlines"
+
+
+class CourseOutlineBackfillAdmin(ReadOnlyAdminMixin, admin.ModelAdmin):
+    """
+    Backfills the course outline for each selected course key.
+    """
+    list_display = ['id']
+    ordering = ['id']
+    search_fields = ['id']
+
+    actions = [backfill_course_outlines_subset, backfill_course_outlines_all]
+
+    def changelist_view(self, request, extra_context=None):
+        """
+        Overrides the admin's changelist_view & selects *all* the CourseOverviews
+        when the custom backfill_course_outlines_all action is selected.
+        """
+        if 'action' in request.POST and request.POST['action'] == 'backfill_course_outlines_all':
```

Comment:
FWIW, I wouldn't worry about reflecting "backfill all" in the UI. There's already a "select all" check box in the standard list view (though that only selects all on the page). I think it's enough for us to have:

1. `backfill_course_outlines_subset` that takes the queryset and launches per-course-key celery tasks directly from the web request.
2. `backfill_course_outlines_all` that just ignores the queryset entirely and launches a celery task that pulls everything. That way we don't really have to care what might happen as the number of courses scales up.

Reply:
Yes - I coded things this way and it's simpler - so it wins!